exonum-java-binding-core/rust/src/storage/core_schema.rs

```diff
@@ -0,0 +1,80 @@
+use exonum::{
+    blockchain::Schema,
+    storage::{Fork, Snapshot, StorageValue},
+};
+use jni::{
+    objects::JClass,
+    sys::{jbyteArray, jlong},
+    JNIEnv,
+};
+use std::{panic, ptr};
+use storage::db::{View, ViewRef};
+use utils::{self, Handle};
+
+type CoreSchema<T> = Schema<T>;
+
+enum SchemaType {
+    SnapshotSchema(CoreSchema<&'static Snapshot>),
+    ForkSchema(CoreSchema<&'static mut Fork>),
+}
+
+/// Returns pointer to created CoreSchemaProxy object
+#[no_mangle]
+pub extern "system" fn Java_com_exonum_binding_blockchain_CoreSchemaProxy_nativeCreate(
+    env: JNIEnv,
+    _: JClass,
+    view_handle: Handle,
+) -> Handle {
+    let res = panic::catch_unwind(|| {
+        let schema_type = match *utils::cast_handle::<View>(view_handle).get() {
+            ViewRef::Snapshot(snapshot) => SchemaType::SnapshotSchema(Schema::new(snapshot)),
+            ViewRef::Fork(ref mut fork) => SchemaType::ForkSchema(Schema::new(fork)),
+        };
+        Ok(utils::to_handle(schema_type))
+    });
+    utils::unwrap_exc_or_default(&env, res)
+}
+
+/// Destroys the underlying `Schema` object and frees memory.
+#[no_mangle]
+pub extern "system" fn Java_com_exonum_binding_blockchain_CoreSchemaProxy_nativeFree(
+    env: JNIEnv,
+    _: JClass,
+    schema_handle: Handle,
+) {
+    utils::drop_handle::<SchemaType>(&env, schema_handle);
+}
+
+/// Returns the height of the latest committed block or -1 if the "genesis block" was not created.
```

Comment:
As far as I read the doc a Panic should be thrown instead of `-1`.
https://docs.rs/exonum/0.8.1/exonum/blockchain/struct.Schema.html#method.height
Let's clarify this point.

Reply:
Yeap, this method throws `java.lang.RuntimeException` in case of panic and returns _error_value_ which is not important for the java side because of exception.